### PR TITLE
Refactor Stream Management in ErizoController

### PR DESCRIPTION
**Pull request by lodoyun**
**Description**
This PR updates #1398 with clean commits, resolved conflicts and a couple of new fixes.

[] It needs and includes Unit Tests

**Changes in Client or Server public APIs**

[] It includes documentation for these changes in `/doc`.

### CODE REVIEW

File: erizo_controller/erizoController/models/Client.js [hunk 1]
```diff
@@ -167,6 +180,9 @@ class Client extends events.EventEmitter {
                          'state: SUBSCRIBER_READY, ' +
                          `streamId: ${signMess.streamId}, ` +
                          `clientId: ${this.id}`);
+        this.roomController
+          .getPublishedStreamById(signMess.streamId)
```
Comment (jcague) [inline erizo_controller/erizoController/models/Client.js]: I'd check if stream is not undefined

File: erizo_controller/erizoController/models/Client.js [hunk 2]
```diff
@@ -176,6 +192,7 @@ class Client extends events.EventEmitter {
         log.error('message: addMultipleSubscribers timeout when contacting ErizoJS, ' +
                           `streamId: ${signMess.streamId}, ` +
                           `clientId: ${this.id}`);
+        this.roomController.getPublishedStreamById(signMess.streamId).removeAvSubscriber(this.id);
```
Comment (jcague) [inline erizo_controller/erizoController/models/Client.js]: I'd check if stream is not undefined

File: erizo_controller/erizoController/models/Client.js [hunk 3]
```diff
@@ -357,6 +382,141 @@ class Client extends events.EventEmitter {
     });
   }
 
+  _publishExternalInput(id, options, sdp, callback) {
```
Comment (jcague) [inline erizo_controller/erizoController/models/Client.js]: I really like all these internal functions!

File: erizo_controller/erizoController/models/Stream.js [hunk 4]
```diff
@@ -2,94 +2,199 @@
 
 /* eslint-disable no-param-reassign */
 
-exports.Stream = (spec) => {
-  const that = {};
-  const dataSubscribers = [];
-
-  that.getID = () => spec.id;
+const StreamStates = Object.freeze({
```
Comment (jcague) [inline erizo_controller/erizoController/models/Stream.js]: 💯

File: erizo_controller/erizoController/roomController.js [hunk 5]
```diff
@@ -239,20 +220,17 @@ exports.RoomController = (spec) => {
...
               log.warn('message: addPublisher ErizoJS timeout no retry, ' +
                                  `retries: ${retries}, streamId: ${streamId}, ` +
                                  `erizoId: ${getErizoQueueFromStreamId(streamId)},`,
                                  logger.objectToLog(options.metadata));
-              const erizo = erizos.findById(publishers[streamId]);
-              if (erizo !== undefined) {
-                const index = erizo.publishers.indexOf(streamId);
-                erizo.publishers.splice(index, 1);
-              }
+              streamManager.updateErizoIdForPublishedStream(streamId, null);
```
Comment (jcague) [inline erizo_controller/erizoController/roomController.js]: Maybe I'm getting this wrong, shouldn't we remove the stream from the list of publishedStreams?

Comment (lodoyun) [reply erizo_controller/erizoController/roomController.js]: We do that in Client.js. The general rule is that roomController only updates the ErizoId for the published streams. The rest (adding/removing publishers, subscribers and update the state) should be done in the client

**Pull request merged**